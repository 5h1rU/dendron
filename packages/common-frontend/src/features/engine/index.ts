export * as engineReducer from "./slice";
export {
  setError,
  setNotes,
  EngineSliceUtils,
  initNotes,
  InitNoteOpts,
  EngineState,
} from "./slice";
export * as engineStore from "./store";
export {
  RootState as engineState,
  AppDispatch as engineDispatch,
} from "./store";
