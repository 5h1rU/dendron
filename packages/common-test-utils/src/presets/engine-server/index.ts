import { ENGINE_DELETE_PRESETS } from "./delete";
import ENGINE_MULTI_TEST_PRESET from "./engine-multi";
import ENGINE_SINGLE_TEST_PRESET from "./engine-single";
import { ENGINE_GET_NOTE_BY_PATH_PRESETS } from "./getByPath";
import { ENGINE_INIT_PRESETS } from "./init";
import NOTE_REF from "./note-refs";
import { ENGINE_QUERY_PRESETS } from "./query";
import { ENGINE_RENAME_PRESETS } from "./rename";
import { ENGINE_UPDATE_PRESETS } from "./update";
import { ENGINE_WRITE_PRESETS } from "./write";

export const ENGINE_SERVER = {
  NOTE_REF,
  ENGINE_MULTI_TEST_PRESET,
  ENGINE_SINGLE_TEST_PRESET,
  ENGINE_WRITE_PRESETS,
  ENGINE_INIT_PRESETS,
  ENGINE_UPDATE_PRESETS,
  ENGINE_DELETE_PRESETS,
  ENGINE_GET_NOTE_BY_PATH_PRESETS,
  ENGINE_RENAME_PRESETS,
  ENGINE_QUERY_PRESETS,
};

export const ENGINE_PRESETS = [
  // {name: "init", presets: ENGINE_SERVER.ENGINE_INIT_PRESETS},
  // {name: "delete", presets: ENGINE_SERVER.ENGINE_DELETE_PRESETS},
  // {name: "getByPath", presets: ENGINE_SERVER.ENGINE_GET_NOTE_BY_PATH_PRESETS},
  { name: "query", presets: ENGINE_SERVER.ENGINE_QUERY_PRESETS },
  // {name: "rename", presets: ENGINE_SERVER.ENGINE_RENAME_PRESETS},
  // {name: "update", presets: ENGINE_SERVER.ENGINE_UPDATE_PRESETS},
  // {name: "write", presets: ENGINE_SERVER.ENGINE_WRITE_PRESETS},
];
