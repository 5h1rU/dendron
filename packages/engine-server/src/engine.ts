import {
  DEngine,
  DEngineStore,
  DNodeData,
  DNodeDict,
  DNodeRawProps,
  EngineQueryResp,
  IDNode,
  INoteOpts,
  NodeBuilder,
  NodeWriteOpts,
  Note,
  NoteDict,
  QueryMode,
  QueryOpts,
  Schema,
  SchemaDict,
  SchemaRawProps,
  Scope,
  assert,
  assertExists,
  createLogger,
  getStage,
  makeResponse
} from "@dendronhq/common-all";

import { BodyParser } from "./drivers/raw/BodyParser";
import FileStorage from "./drivers/file/store";
import Fuse from "fuse.js";
import _ from "lodash";
import fs from "fs-extra";

let PROTO_ENGINE: ProtoEngine;
const logger = createLogger("DEngine");

function isAllQuery(qs: string): boolean {
  return qs === "**/*";
}

export interface FuseOptions {
  exactMatch?: boolean;
  preset: QueryMode;
  // isCaseSensitive?: boolean
  // distance?: number
  // findAllMatches?: boolean
  // includeMatches?: boolean
  // includeScore?: boolean
  // location?: number
  // minMatchCharLength?: number
  // shouldSort?: boolean
  // threshold?: number
  // useExtendedSearch?: boolean
}

function createFuse<T>(initList: T[], opts: FuseOptions) {
  const options = {
    shouldSort: true,
    threshold: opts.exactMatch ? 0.0 : 0.6,
    location: 0,
    distance: 100,
    maxPatternLength: 32,
    minMatchCharLength: 1,
    keys: ["title", "path"]
  };
  if (opts.preset === "schema") {
    options.keys = ["title", "id"];
  }
  // initList = _.map(initList, (n) => ({ ...n, treePath: n.path }));
  const fuse = new Fuse(initList, options);
  return fuse;
}

type ProtoEngineOpts = {
  root?: string;
  cacheDir?: string;
  forceNew?: boolean;
};

type ProtoEngineProps = Required<ProtoEngineOpts>;

export class ProtoEngine implements DEngine {
  public fuse: Fuse<Note, any>;
  public schemaFuse: Fuse<Schema, any>;

  public notes: NoteDict;
  public schemas: SchemaDict;

  public fullNodes: Set<string>;

  public opts: ProtoEngineProps;

  public queries: Set<string>;

  public store: DEngineStore;

  static getEngine(opts?: ProtoEngineOpts): DEngine {
    if (!PROTO_ENGINE || opts?.forceNew) {
      const stage = getStage();
      opts = _.defaults(opts || {}, {
        // TODO: remove hardcoded value
        root: "/Users/kevinlin/Dropbox/Apps/Noah/notesv2"
      });
      if (_.isUndefined(opts.root)) {
        throw Error(`root must be defined`);
      }
      // TODO
      PROTO_ENGINE = new ProtoEngine(
        new FileStorage(opts as { root: string }),
        opts
      );
      logger.info({ ctx: "getEngine:exit", opts, stage });
      return PROTO_ENGINE;
    }
    return PROTO_ENGINE;
  }

  constructor(store: DEngineStore, opts: ProtoEngineOpts) {
    // eslint-disable-next-line spaced-comment
    //this.nodes = store.fetchInitial();
    this.notes = {};
    this.store = store;
    this.fuse = createFuse<Note>([], { exactMatch: false, preset: "note" });
    this.fullNodes = new Set();
    this.queries = new Set();

    // setup schemas
    this.schemas = {};
    this.schemaFuse = createFuse<Schema>([], {
      exactMatch: false,
      preset: "schema"
    });

    this.opts = _.defaults(opts, {
      cacheDir: "/tmp/dendronCache",
      root: "/Users/kevinlin/Dropbox/Apps/Dendron",
      forceNew: false
    });
    [opts.cacheDir, opts.root].forEach(fpath => {
      if (!fs.existsSync(fpath as string)) {
        throw Error(`${fpath} doesn't exist`);
      }
    });
  }

  deleteFromNodes(id: string) {
    delete this.notes[id];
    this.fullNodes.delete(id);
  }

  /**
   * Updates local cache
   * - if node is new, add it to cache
   * - if not new, update existing node properties
   * - extra:
   *   - if full node, add id
   * @param nodes
   * @param opts
   */
  refreshNodes(nodes: IDNode[], opts?: QueryOpts) {
    if (_.isEmpty(nodes)) {
      return;
    }
    const mode = nodes[0].type;
    if (mode === "schema") {
      const schemas = nodes as Schema[];
      schemas.forEach((node: Schema) => {
        this.schemas[node.id] = node;
      });
      this.updateLocalCollection(_.values(this.schemas), "schema");
      return;
    } else {
      nodes.forEach((node: IDNode) => {
        const { id } = node;

        if (!_.has(this.notes, id)) {
          // add if not exist
          // TODO: nodes has both raw and full nodes
          // @ts-ignore
          this.notes[id] = node;
        } else {
          // exists, merge it
          logger.debug({
            ctx: "refreshNodes:existingNode",
            node: node.toRawProps()
          });
          _.merge(this.notes[id], node);
        }
        // if a full node, make sure to add
        if (opts?.fullNode) {
          this.fullNodes.add(id);
        }
      });
      // FIXME: debug
      const newNodes = _.map(nodes, n => ({ title: n.title, id: n.id }));
      const allNodes = _.map(this.notes, n => ({
        title: n.title,
        id: n.id
      }));
      logger.debug({
        ctx: "refreshNodes",
        newNodes,
        allNodes
      });
      this.updateLocalCollection(_.values(this.notes), "note");
      return;
    }
  }

  /**
   * Turn parent|children ids to full nodes
   */
  async resolveIds(
    node: DNodeRawProps<DNodeData>,
    dict: DNodeDict
  ): Promise<Note> {
    const nodeResolved: INoteOpts = dict[node.id] as INoteOpts;
    assert(
      !_.isUndefined(nodeResolved),
      `node ${JSON.stringify(node)} not in dict\nkeys: ${_.keys(dict)}`
    );
    nodeResolved.body = node.body;
    return new Note({ ...nodeResolved });
  }

  updateLocalCollection(collection: IDNode[], mode: QueryMode) {
    if (mode === "schema") {
      return this.schemaFuse.setCollection(collection as Schema[]);
    } else {
      return this.fuse.setCollection(collection as Note[]);
    }
  }

  async delete(id: string): Promise<void> {
    await this.store.delete(id);
    this.deleteFromNodes(id);
    return;
  }

  /**
   * Defaults
   * @param id
   * @param opts
   *  - fullNode: {default:true}
   */
  async get(_scope: Scope, id: string, mode: QueryMode, opts?: QueryOpts) {
    opts = _.defaults(opts || {}, { fullNode: true, createIfNew: true });
    let nodeDict;
    logger.debug({ ctx: "get", id, opts });

    if (mode === "schema") {
      nodeDict = this.schemas;
      const node = nodeDict[id];
      return { data: node };
    }
    //FIXME: check if exist
    // we are assuming all nodes are fetched
    const node = this.notes[id];

    // a full node has a body and is fully resolved
    if (opts?.fullNode && !this.fullNodes.has(id)) {
      logger.debug({ ctx: "get:fetchFromStore:pre", id });
      const fnResp = await this.store.get(_scope, id, {
        ...opts,
        webClient: true
      });
      logger.debug({ ctx: "get:fetchFromStore:post", id, opts, fnResp });
      const fullNode = await this.resolveIds(fnResp.data, this.notes);
      logger.debug({ ctx: "get:resolve:post", fnResp });
      // TODO:
      this.refreshNodes([fullNode], opts);
      return { data: fullNode };
    }
    logger.debug({ ctx: "get:exit", node });
    return { data: node };
  }

  async query(
    scope: Scope,
    queryString: string,
    mode: QueryMode,
    opts?: QueryOpts
  ) {
    opts = _.defaults(opts || {}, {
      fullNode: false,
      createIfNew: true,
      initialQuery: false
    });
    let data: EngineQueryResp;

    // handle all query case
    if (isAllQuery(queryString)) {
      logger.debug({ ctx: "query:queryAll:pre", mode });
      data = await this.store.query(scope, "**/*", mode, opts);
      if (opts.initialQuery) {
        this.refreshNodes(data.data);
      }
      // @ts-ignore
      return data as EngineQueryResp;
    }

    // handle schema query
    if (mode === "schema") {
      const results = this.schemaFuse.search(queryString);
      let items: Schema[];
      const nodes = this.schemas;
      if (opts.queryOne) {
        items = [results[0]?.item];
      } else {
        items = _.map(results, resp => resp.item);
      }
      logger.debug({ ctx: "query:exit:schema", items });
      return makeResponse<EngineQueryResp>({
        data: _.map(items, item => nodes[item.id]),
        error: null
      });
    } else {
      // handle note query
      const results = this.fuse.search(queryString);
      let items: IDNode[];
      if (opts.queryOne) {
        items = [results[0]?.item];
      } else {
        items = _.map(results, resp => resp.item);
      }

      if (opts.queryOne && items[0]?.path !== queryString && opts.createIfNew) {
        logger.debug({
          ctx: "query:write:pre",
          queryString,
          item: items[0]
        });
        const nodeBlank = new Note({ fname: queryString });
        await this.write(scope, nodeBlank, { newNode: true });
        const { data: nodeFull } = await this.get(scope, nodeBlank.id, mode);
        this.refreshNodes([nodeFull], { fullNode: true });
        return makeResponse<EngineQueryResp>({
          data: [nodeFull],
          error: null
        });
      }

      if (opts.fullNode) {
        const fetchedFullNodes = await Promise.all(
          _.map<IDNode, Promise<IDNode | null>>(items, async ent => {
            if (!this.fullNodes.has(ent.id)) {
              logger.debug({
                ctx: "query:fuse.search:post",
                status: "fetch full node from store"
              });
              // FIXME: ratelimit
              const fn = await this.get(scope, ent.id, mode);
              return fn.data;
            }
            logger.debug({
              ctx: "query:fuse.search:post",
              status: "fetch full node from cache"
            });
            return null;
          })
        );
        this.refreshNodes(
          _.filter(fetchedFullNodes, ent => !_.isNull(ent)) as IDNode[],
          { fullNode: true }
        );
        logger.debug({
          ctx: "query:fetchedFullNodes:exit",
          fetchedFullNodes
        });
      }
      logger.debug({ ctx: "query:exit:note", items });
      return makeResponse<EngineQueryResp>({
        data: _.map(items, item => this.notes[item.id]),
        error: null
      });
    }
  }

  async write(scope: Scope, node: IDNode, opts?: NodeWriteOpts): Promise<void> {
    opts = _.defaults(opts, { newNode: false, body: "" });
    if (node.type === "schema") {
      assertExists(opts.body, "body must exist");
      // convert body
      const props: SchemaRawProps[] = new BodyParser().parseSchema(
        opts.body as string,
        {
          node,
          fname: node.fname
        }
      );
      const schemas = new NodeBuilder().buildSchemaFromProps(props);
      const schemaNew = _.find(schemas, { id: node.id });
      if (_.isUndefined(schemaNew)) {
        throw Error(`no schema found for ${node} on write`);
      }
      node = schemaNew;
      // reset body
      node.body = "";
      await this.store.write(scope, node);
      if (opts.newNode) {
        const parentPath = "root";
        const parentNode = _.find(this.schemas, n => n.title === parentPath);
        if (!parentNode) {
          throw Error("no parent found");
        }
        (parentNode as Schema).addChild(node as Schema);
      }
    } else {
      const note = node as Note;
      await this.store.write(scope, note);
      if (opts.newNode) {
        let parentPath = note.fname
          .split(".")
          .slice(0, -1)
          .join(".");
        if (_.isEmpty(parentPath)) {
          parentPath = "root";
        }
        const parentNode = _.find(this.notes, n => n.path === parentPath);
        if (!parentNode) {
          throw Error("no parent found");
        }
        parentNode.addChild(note);
      }
    }
    // TODO
    return this.refreshNodes([node], { fullNode: opts.newNode });
  }
}

export function engine() {
  return ProtoEngine.getEngine();
}
