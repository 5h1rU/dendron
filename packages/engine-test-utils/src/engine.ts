import { DEngineClientV2, DVault, WorkspaceOpts } from "@dendronhq/common-all";
import { launch } from "@dendronhq/api-server";
import { tmpDir } from "@dendronhq/common-server";
import {
  RunEngineTestFunctionV4,
  runJestHarnessV2,
  SetupHookFunction,
} from "@dendronhq/common-test-utils";
import {
  createEngine as engineServerCreateEngine,
  DendronEngineClient,
  WorkspaceService,
} from "@dendronhq/engine-server";
import _ from "lodash";

export type AsyncCreateEngineFunction = (
  opts: WorkspaceOpts
) => Promise<DEngineClientV2>;

export async function createEngineFromEngine(opts: WorkspaceOpts) {
  return engineServerCreateEngine(opts);
}

export { DEngineClientV2, DVault, WorkspaceOpts };

export async function createEngineFromServer(opts: WorkspaceOpts) {
  const port = await launch({});
  const engine: DEngineClientV2 = DendronEngineClient.create({
    port,
    vaultsv4: opts.vaults,
    ws: opts.wsRoot,
    vaults: [],
  });
  await engine.init();
  return engine;
}

async function setupWS(opts: { vaults: DVault[] }) {
  const wsRoot = tmpDir().name;
  const ws = new WorkspaceService({ wsRoot });
  const vaults = await Promise.all(
    opts.vaults.map(async (vault) => {
      await ws.createVault({ vault });
      return vault;
    })
  );
  return { wsRoot, vaults };
}

// @ts-ignore
export async function runEngineTestV5<TExtra = any>(
  func: RunEngineTestFunctionV4,
  opts: {
    preSetupHook?: SetupHookFunction;
    //postSetupHook?: PostSetupHookFunction;
    createEngine?: AsyncCreateEngineFunction;
    extra?: TExtra;
    expect: any;
    vaults?: DVault[];
    setupOnly?: boolean;
  }
): Promise<any> {
  const { preSetupHook, extra, vaults, createEngine } = _.defaults(opts, {
    preSetupHook: async ({}) => {},
    postSetupHook: async ({}) => {},
    createEngine: createEngineFromEngine,
    extra: {},
    vaults: [{ fsPath: "vault1" }, { fsPath: "vault2" }],
  });
  const { wsRoot } = await setupWS({ vaults });
  await preSetupHook({ wsRoot, vaults });
  const engine = await createEngine({ wsRoot, vaults });
  const initResp = await engine.init();
  const testOpts = { wsRoot, vaults, engine, initResp, extra };
  if (opts.setupOnly) {
    return testOpts;
  }
  const results = (await func(testOpts)) || [];
  await runJestHarnessV2(results, expect);
  return { opts: testOpts, resp: undefined };
}
