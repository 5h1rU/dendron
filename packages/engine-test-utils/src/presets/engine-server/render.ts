import { TestPresetEntryV4 } from "@dendronhq/common-test-utils";
import { ENGINE_HOOKS } from "./utils";

const NOTES = {
  BASIC: new TestPresetEntryV4(
    async ({ engine }) => {
      const { data } = await engine.renderNote({
        id: "foo",
        format: "markdown",
      });
      return [{ actual: data, expected: "<p>foo body</p>", msg: "foo" }];
    },
    {
      preSetupHook: async (opts) => {
        return ENGINE_HOOKS.setupBasic(opts);
      },
    }
  ),
};

export const ENGINE_RENDER_PRESETS = {
  NOTES,
};
