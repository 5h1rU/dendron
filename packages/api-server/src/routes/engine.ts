import {
  DEngineV2,
  EngineQueryRequest,
  EngineWriteRequest,
} from "@dendronhq/common-all";
import { DendronEngine } from "@dendronhq/engine-server";
import { Request, Response, Router } from "express";
import { MemoryStore } from "../store/memoryStore";

const router = Router();

router.get("/health", async (_req: Request, res: Response) => {
  return res.json({ ok: 1 });
});

router.post("/query", async (req: Request, res: Response) => {
  const { ws, queryString, mode } = req.body as EngineQueryRequest;
  const engine = await MemoryStore.instance().get<DendronEngine>(`ws:${ws}`);
  if (!engine) {
    throw "No Engine";
  }
  const { error, data } = await engine.query(queryString, mode);
  const cleanData = data.map((ent) => ent.toRawProps());
  res.json({ error, data: cleanData });
});

router.post("/write", async (req: Request, res: Response) => {
  const { ws, node, opts } = req.body as EngineWriteRequest;
  const engine = await MemoryStore.instance().get<DEngineV2>(`ws:${ws}`);
  if (!engine) {
    throw "No Engine";
  }
  try {
    await engine.writeNote(node, opts);
    res.json({ error: null, data: null });
  } catch (err) {
    res.json({ error: JSON.stringify(err), data: null });
  }
});

export { router as engineRouter };
