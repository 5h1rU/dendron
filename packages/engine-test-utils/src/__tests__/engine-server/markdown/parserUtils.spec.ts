import { DNoteLoc } from "@dendronhq/common-all";
import { NoteTestUtilsV4 } from "@dendronhq/common-test-utils";
import { ParserUtilsV2 } from "@dendronhq/engine-server";
import _ from "lodash";
import { runEngineTestV5, testWithEngine } from "../../../engine";

describe("parserUtils", () => {
  describe("findHeaders", async () => {
    test("one header", async () => {
      await runEngineTestV5(
        async ({ engine }) => {
          const body = engine.notes["foo"].body;
          const out = ParserUtilsV2.findHeaders(body);
          expect(out).toMatchSnapshot("bond");
          expect(_.size(out)).toEqual(1);
          expect(out[0].depth).toEqual(1);
        },
        {
          preSetupHook: async ({ vaults, wsRoot }) => {
            await NoteTestUtilsV4.createNote({
              fname: "foo",
              body: "# h1",
              vault: vaults[0],
              wsRoot,
            });
          },
          expect,
        }
      );
    });
  });

  describe("findLinks", async () => {
    testWithEngine(
      "one link",
      async ({ engine }) => {
        const note = engine.notes["foo"];
        const links = ParserUtilsV2.findLinks({ note, engine });
        expect(links).toMatchSnapshot("bond");
        expect(links[0].to?.fname).toEqual("bar");
      },
      {
        preSetupHook: async ({ wsRoot, vaults }) => {
          await NoteTestUtilsV4.createNote({
            fname: "foo",
            body: "[[bar]]",
            vault: vaults[0],
            wsRoot,
          });
        },
      }
    );

    testWithEngine(
      "empty link",
      async ({ engine }) => {
        const note = engine.notes["foo"];
        const links = ParserUtilsV2.findLinks({ note, engine });
        expect(links).toMatchSnapshot();
        expect(_.isEmpty(links)).toBeTruthy();
      },
      {
        preSetupHook: async ({ wsRoot, vaults }) => {
          await NoteTestUtilsV4.createNote({
            fname: "foo",
            body: "[[]]",
            vault: vaults[0],
            wsRoot,
          });
        },
      }
    );

    testWithEngine(
      "xvault link",
      async ({ engine }) => {
        const note = engine.notes["foo"];
        const links = ParserUtilsV2.findLinks({ note, engine });
        expect(links).toMatchSnapshot();
        expect(links[0].from).toEqual({
          fname: "foo",
          id: "foo",
          vault: {
            fsPath: "vault1",
          },
        });
        expect(links[0].to).toEqual({
          fname: "bar",
          vault: {
            fsPath: "vault2",
          },
        });
      },
      {
        preSetupHook: async ({ wsRoot, vaults }) => {
          await NoteTestUtilsV4.createNote({
            fname: "foo",
            body: "[[dendron://vault2/bar]]",
            vault: vaults[0],
            wsRoot,
          });
        },
      }
    );
  });

  describe("replaceLinks", async () => {
    testWithEngine("one", async () => {
      const content = ["[[bond]]"].join("\n");
      const from: DNoteLoc = {
        fname: "bond",
        vault: {
          fsPath: "/tmp",
        },
      };
      const to: DNoteLoc = {
        fname: "bar",
        vault: {
          fsPath: "/tmp",
        },
      };
      const out = await ParserUtilsV2.replaceLinks({ content, from, to });
      expect(out).toMatchSnapshot();
      expect(_.trim(out)).toEqual("[[bar]]");
    });

    testWithEngine("ref", async ({ vaults }) => {
      const vault = vaults[0];
      const content = ["![[bond]]"].join("\n");
      const from: DNoteLoc = {
        fname: "bond",
        vault,
      };
      const to: DNoteLoc = {
        fname: "bar",
        vault,
      };
      const out = await ParserUtilsV2.replaceLinks({ content, from, to });
      expect(out).toMatchSnapshot();
      expect(_.trim(out)).toEqual("![[bar]]");
    });

    testWithEngine("multiple", async ({ vaults }) => {
      const vault = vaults[0];
      const content = ["[[bond]]", "[[bond]]"].join("\n");
      const from: DNoteLoc = {
        fname: "bond",
        vault,
      };
      const to: DNoteLoc = {
        fname: "bar",
        vault,
      };
      const out = await ParserUtilsV2.replaceLinks({ content, from, to });
      expect(out).toMatchSnapshot();
    });

    testWithEngine("inline code", async ({ vaults }) => {
      const vault = vaults[0];
      const content = ["`[[bond]]`"].join("\n");
      const from: DNoteLoc = {
        fname: "bond",
        vault,
      };
      const to: DNoteLoc = {
        fname: "bar",
        vault,
      };
      const out = await ParserUtilsV2.replaceLinks({ content, from, to });
      expect(out).toMatchSnapshot();
      expect(_.trim(out)).toEqual("`[[bond]]`");
    });

    testWithEngine("fenced code", async ({ vaults }) => {
      const vault = vaults[0];
      const content = ["```", "[[bond]]", "```"].join("\n");
      const from: DNoteLoc = {
        fname: "bond",
        vault,
      };
      const to: DNoteLoc = {
        fname: "bar",
        vault,
      };
      const out = await ParserUtilsV2.replaceLinks({ content, from, to });
      expect(out).toMatchSnapshot();
    });

    testWithEngine("with alias", async ({ vaults }) => {
      const content = ["[[hero|bond]]"].join("\n");
      const vault = vaults[0];
      const from: DNoteLoc = {
        fname: "bond",
        vault,
      };
      const to: DNoteLoc = {
        fname: "bar",
        vault,
      };
      const out = await ParserUtilsV2.replaceLinks({ content, from, to });
      expect(out).toMatchSnapshot();
      expect(_.trim(out)).toEqual("[[hero|bar]]");
    });

    testWithEngine("with offset ", async ({ vaults }) => {
      const vault = vaults[0];
      const content = ["   [[bond]]"].join("\n");
      const from: DNoteLoc = {
        fname: "bond",
        vault,
      };
      const to: DNoteLoc = {
        fname: "bar",
        vault,
      };
      const out = await ParserUtilsV2.replaceLinks({ content, from, to });
      expect(out).toMatchSnapshot();
      expect(_.trim(out)).toEqual("[[bar]]");
    });
  });
});
