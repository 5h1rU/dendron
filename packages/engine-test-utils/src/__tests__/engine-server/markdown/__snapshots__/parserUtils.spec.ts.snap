// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parserUtils findHeaders one header: bond 1`] = `
Array [
  Object {
    "children": Array [
      Object {
        "position": Position {
          "end": Object {
            "column": 5,
            "line": 1,
            "offset": 4,
          },
          "indent": Array [],
          "start": Object {
            "column": 3,
            "line": 1,
            "offset": 2,
          },
        },
        "type": "text",
        "value": "h1",
      },
    ],
    "depth": 1,
    "position": Position {
      "end": Object {
        "column": 5,
        "line": 1,
        "offset": 4,
      },
      "indent": Array [],
      "start": Object {
        "column": 1,
        "line": 1,
        "offset": 0,
      },
    },
    "type": "heading",
  },
]
`;

exports[`parserUtils findLinks empty link 1`] = `Array []`;

exports[`parserUtils findLinks one link: bond 1`] = `
Array [
  Object {
    "alias": "bar",
    "from": Object {
      "fname": "foo",
      "id": "foo",
      "vault": Object {
        "fsPath": "vault1",
      },
    },
    "original": "bar",
    "pos": Object {
      "end": 7,
      "start": 0,
    },
    "to": Object {
      "anchorHeader": undefined,
      "fname": "bar",
      "vault": undefined,
    },
    "type": "wiki",
    "value": "bar",
  },
]
`;

exports[`parserUtils findLinks xvault link 1`] = `
Array [
  Object {
    "alias": "bar",
    "from": Object {
      "fname": "foo",
      "id": "foo",
      "vault": Object {
        "fsPath": "vault1",
      },
    },
    "original": "bar",
    "pos": Object {
      "end": 24,
      "start": 0,
    },
    "to": Object {
      "anchorHeader": undefined,
      "fname": "bar",
      "vault": Object {
        "fsPath": "vault2",
      },
    },
    "type": "wiki",
    "value": "bar",
  },
]
`;

exports[`parserUtils replaceLinks fenced code 1`] = `
"\`\`\`
[[bond]]
\`\`\`

"
`;

exports[`parserUtils replaceLinks inline code 1`] = `
"\`[[bond]]\`

"
`;

exports[`parserUtils replaceLinks multiple 1`] = `
"[[bar]]
[[bar]]

"
`;

exports[`parserUtils replaceLinks one 1`] = `
"[[bar]]

"
`;

exports[`parserUtils replaceLinks ref 1`] = `
"![[bar]]

"
`;

exports[`parserUtils replaceLinks with alias 1`] = `
"[[hero|bar]]

"
`;

exports[`parserUtils replaceLinks with offset  1`] = `
"   [[bar]]

"
`;
